### PR TITLE
459 outputs structure

### PR DESCRIPTION
**Pull request by zorge69**
## Pull Request submission
Check the files:
pipeline.py - it should run one script, run_outputs from outputs_main.py
outputs_main.py - it has a single, conditional entry for each output type, at present, short_form and tau
short_form,py - does everythin for short form
tau.py  - work in progress
map_output_cols.py - added one function to add pg numeric
output_helpers.py - other reusable functions that can help other output modules; presently, just one function that extracts data and applies an output schema.

Run the feature setting short form output True, tau output False - it should work.

*Insert any instructions to help the reviewer, e.g. "install new requirements from `environment.yml`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [x] All requirements install from (updated) `environment.yaml`
- [x] Documentation has been created and is clear - **check the working document**
- [x] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [x] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [x] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [x] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/outputs/outputs_helpers.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    # Rename columns to match the output specification
+    output_df.rename(
+        columns={key: colname_dict[key] for key in colname_dict}, inplace=True
+    )
+
+    # Rearrange to match user defined output order
+    output_df = output_df[colname_dict.values()]
+
+    # Validate datatypes before output
+    validate_data_with_schema(output_df, schema)
```
Comment (dpurches) [inline src/outputs/outputs_helpers.py]: I think this should work, but may be inefficient as I believe this function not only validates but also enforces column types in pandas if they are wrong. If we are giving an output with out types like a .csv this obviously wouldn't be necessary.

File: src/outputs/outputs_main.py [hunk 2]
```diff
@@ -8,14 +8,21 @@
...
+from src.outputs.tau import run_tau
+
+
 OutputMainLogger = logging.getLogger(__name__)
 
 # Get the shortform schema
 short_form_schema = toml.load("src/outputs/output_schemas/frozen_shortform_schema.toml")
 
+# Get the Tau Argus output schema
+tau_schema = toml.load("src/outputs/output_schemas/tau_schema.toml")
```
Comment (dpurches) [inline src/outputs/outputs_main.py]: Should be done inside the function maybe?

File: src/outputs/short_form.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+from src.staging.validation import load_schema, validate_data_with_schema
+from src.outputs.outputs_helpers import create_output_df
+
+
+OutputMainLogger = logging.getLogger(__name__)
+
+# Get the shortform schema
+short_form_schema = toml.load("src/outputs/output_schemas/frozen_shortform_schema.toml")
+
+def run_short_form(
```
Comment (dpurches) [inline src/outputs/short_form.py]:
```suggestion
def output_short_form(
```
Might be more specific

File: src/outputs/short_form_out.py
Comment (dpurches) [inline src/outputs/short_form_out.py:1]: I'm guessing we'll just have either `short_form.py` or `short_form_out.py` in the finished version?

File: src/outputs/output_schemas/tau_schema.toml
Comment (dpurches) [inline src/outputs/output_schemas/tau_schema.toml:1]: Might be a pain to sort out, but in the long run I think it would be good to have these in the `config` folder and have separate `config/inputs` and `config/outputs` folders.

File: src/outputs/outputs_helpers.py (hunk 1 above)
Comment (zorge69) [reply src/outputs/outputs_helpers.py]: Agree, removed validation and dtype enforcement.

File: src/outputs/outputs_main.py (hunk 2 above)
Comment (zorge69) [reply src/outputs/outputs_main.py]: Agree, it should be read inside the function from the config

File: src/outputs/short_form.py (hunk 3 above)
Comment (zorge69) [reply src/outputs/short_form.py]: agree, done

File: src/outputs/short_form_out.py
Comment (zorge69) [reply src/outputs/short_form_out.py:1]: yes, done

File: tests/test_outputs/test_outputs_helpers.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from msilib import schema
```
Comment (JenCheshire) [inline tests/test_outputs/test_outputs_helpers.py]: I think this line can be taken out - it doesn't look like it's needed (tests run fine when I comment it out), and I think it's the line causing the checks to fail in the PR

Comment (zorge69) [reply tests/test_outputs/test_outputs_helpers.py]: Yes, thanks - that worked!

File: src/estimation/estimation_main.py [hunk 5]
```diff
@@ -46,9 +46,21 @@ def run_estimation(
     # calculate the weights for outliers
     weighted_df = weights.outlier_weights(df)
 
+    # temporary - for debugging
```
Comment (JenCheshire) [inline src/estimation/estimation_main.py]: To remove?

File: src/estimation/estimation_main.py [hunk 6]
```diff
@@ -46,9 +46,21 @@ def run_estimation(
...
+    out_fol = r"D:\data\res_dev\estimation"
+    mypath = os.path.join(out_fol, "weighted_df.csv")
+    weighted_df.to_csv(mypath, index=None)
+    # end of debugging
+
     # apply the weights to the dataframe and apply the specified rounding
     estimated_df = appweights.apply_weights(weighted_df, 4)
-
+    
+    # temporary - for debugging
```
Comment (JenCheshire) [inline src/estimation/estimation_main.py]: To remove?

File: join_pg_prototype.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
Comment (JenCheshire) [inline join_pg_prototype.py]: Are these files still required?

File: pg_alpha_num.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
```
Comment (JenCheshire) [inline pg_alpha_num.py]: Are these files still required?

**Review by JenCheshire — COMMENTED** [review]
Looking good.  Just a couple of questions about two of the files, and some temp code

File: src/outputs/tau.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    """Run the outputs module.
+
+    Args:
+        df (pd.DataFrame): The dataset main with weights not applied
+        config (dict): The configuration settings.
+        write_csv (Callable): Function to write to a csv file.
+         This will be the hdfs or network version depending on settings.
+        run_id (int): The current run id
+        ultfoc_mapper (pd.DataFrame): The ULTFOC mapper DataFrame.
+        cora_mapper (pd.DataFrame): used for adding cora "form_status" column
```
Comment (JenCheshire) [inline src/outputs/tau.py:34]: postcode_itl_mapper missing from the docstring

File: src/outputs/output_schemas/tau_schema.toml
Comment (zorge69) [reply src/outputs/output_schemas/tau_schema.toml:1]: Nice to have, fix some other time?

File: src/estimation/estimation_main.py (hunk 5 above)
Comment (zorge69) [reply src/estimation/estimation_main.py]: Removed the code snippet

File: src/estimation/estimation_main.py (hunk 6 above)
Comment (zorge69) [reply src/estimation/estimation_main.py]: Removed the piece of code

File: join_pg_prototype.py (hunk 7 above)
Comment (zorge69) [reply join_pg_prototype.py]: removed the entire file join_pg_prptptype.py

File: pg_alpha_num.py (hunk 8 above)
Comment (zorge69) [reply pg_alpha_num.py]: removed the entire file pg_alpha_num.py

File: src/outputs/outputs_main.py (hunk 2 above)
Comment (dpurches) [reply src/outputs/outputs_main.py]: I don't think this has been actioned yet

File: src/outputs/output_schemas/tau_schema.toml
Comment (dpurches) [reply src/outputs/output_schemas/tau_schema.toml:1]: Yeh I agree, will resolve

File: src/outputs/outputs_main.py (hunk 2 above)
Comment (zorge69) [reply src/outputs/outputs_main.py]: Removed loading schemas in outputs_main.py. They are called from within corresponding modules. Also removed import toml

**Review by JenCheshire — APPROVED** [review]
Looks great
